### ISSUE TITLE
Workspace OOM & OOD Notifications

### ISSUE THREAD

**Issue by defelmnq** [coder/coder#15746]
### Problem statement

Users commonly suffer from crashed/disconnected workspaces due to resource constraints on the agent. The most common of these issues is OOM (out-of-memory) / OOD (out-of-disk) errors that disconnect users while they're working. Users want to know ahead of time when they are running up against the limits of their workspace's allocated resources. 

Both users and administrators want to diagnose whether the agent was killed due to OOM/OOD _after_ crashes occur. 

### Proposal

To remedy this, we're adding a native OOM/OOD notification for workspaces as part of a global initiative to improve the notifications feature.

This feature - in a few words - allow users to set threshold and receive notifications is either the Memory or Disk resource of their workspace is reached.

---

In term of modifications, it requires : 

coder/internal#246 
- Some changes on the Agent to fetch the data
- Update of the AgentAPI to push the fetched data to the CP

coder/internal#245 
- Update of templates to include the thresholds for both memory and disk.
- Have the threshold stored in the DB and retrievable.

coder/internal#248
- Create a piece of logic processing the data points sent by the Agents
- Include a debouncing logic in it to avoid spamming notifications

coder/internal#247 
- Add a migration in the notifiers to send this new kind of email
- Modify both the FE and BE to have a new event configurable under `Workspace events`.

**Comment by defelmnq** [comment]
Some points for later remains : 

- Do we want to have a hardcoded value for the debouncing - value used to ignore notifications when a first one has been sent // which value ? 
- Do we want to store anything for historical purpose ? The alerts that have been triggered, or any data historical points (for example 1d or something.)
- As discussed with @mtojek - a workspace can have multiple agents , we need to keep that in mind while notifying the user. Should these agents work in a completely isolated way , with potentially 2 or more agents triggering alerts at the same time ?
- @stirby @dannykopping (as I know you already raised this point) - Should we let the users define which partition of the disk they want to monitor ? Spike also has some opinions here.

**Comment by mtojek** [comment]
I'll try to answer some questions to unblock you:

> Do we want to have a hardcoded value for the debouncing

Let's hardcode them for now, we can always expose them via configuration later.

> Do we want to store anything for historical purpose ? 

Good question, deferring to @stirby 

> Should these agents work in a completely isolated way , with potentially 2 or more agents triggering alerts at the same time

I would say so. Multiple agents is not a common use case, so I doubt that somebody will boot 10 agents and have all of them OOMed :)

**Comment by stirby** [comment]
> Do we want to store anything for historical purpose ?

Great question; users will definitely want to see these alerts after the fact. Since users are getting this via SMTP in most cases, they can retro in their inbox. However, administrators may want to see these events in a log so they can diagnose bulk OOM. 

"Capturing reason for disconnect" has been a common request, it'd be good to capture this historically.


> Should we let the users define which partition of the disk they want to monitor? Spike also has some opinions here.
 
Yes. Some Terraform examples may help us imagine the UX.

**Comment by dannykopping** [comment]
From the description:

> Both users and administrators want to diagnose whether the agent was killed due to OOM/OOD _after_ crashes occur.

Strictly speaking this is out of scope. Just because an OOM notification was received before a crash doesn't _necessarily_ mean this was the cause. To actually allow historical analysis we'll have to do more than just notifications. I'd suggest removing this from the description.

> _Need to be done later_
> 
> * Create a piece of logic processing the data points sent by the Agents
> * Include a debouncing logic in it to avoid spamming notifications

I don't understand; this is core to this feature; why would it be done later? We can't merge this feature without debouncing; it'd be far too noisy.

> _Need to be done later_
> 
> * Add a migration in the notifiers to send this new kind of email
> * Modify both the FE and BE to have a new event configurable under `Workspace events`

Ditto; this is part-and-parcel of this feature so we should just create sub-issues for them now.

> * Do we want to store anything for historical purpose ? The alerts that have been triggered, or any data historical points (for example 1d or something.)

We'll have records of these notifications in the `notification_messages` table for 7d before they're purged, and as @stirby correctly points out - a user's mailbox is already an archive of these. If customers ask for more specific observability we could always create a metric which tracks these; this might be good to do in general to track the frequency of certain notifications.

> * As discussed with [@mtojek](https://github.com/mtojek) - a workspace can have multiple agents , we need to keep that in mind while notifying the user. Should these agents work in a completely isolated way , with potentially 2 or more agents triggering alerts at the same time ?

I'm thinking we can link the monitoring to the agent, perhaps. See Terraform proposal below:

---

Here's my idea for how the monitoring could be implemented on the template-level:

```hcl
module "coder-monitoring" {
  source   = "registry.coder.com/modules/coder-monitoring/coder"
  version  = ">= 1.0.0"
  agent_id = coder_agent.dev.id # link to agent
  
  threshold = 90 # percent threshold for all, unless overridden by *_threshold
  memory_threshold = 95
  disk_threshold = data.coder_parameter.disk_alert_threshold.value  # could allow user-level
  disks = ["/", "/home/${data.coder_workspace_owner.me.name}"]

  enabled = data.coder_parameter.alert_on_low_resources.value  # could allow user-level control
}
```

**Comment by defelmnq** [comment]
Small update here - First thanks everyone for the different answers. Really help to scope and ensure we're going on the right direction 🙏 

I started the implementation based on the proposal above from Danny , seems this one really gives us the flexibility we want while staying simple enough.

**Comment by ericpaulsen** [comment]
> As discussed with [@mtojek](https://github.com/mtojek) - a workspace can have multiple agents , we need to keep that in mind while notifying the user.

my perspective from the field - we do not typically see customers leveraging multiple agents per workspace. while this workflow is technically supported, I wouldn't spend too many cycles solving for this as it is not common.

**Comment by mtojek** [comment]
That is correct although we have to be mindful while designing it. We can't draft a solution that can't support multiple agents by design as this will increase the technical debt.

**Comment by stirby** [comment]
> > Both users and administrators want to diagnose whether the agent was killed due to OOM/OOD after crashes occur.

> Strictly speaking this is out of scope. Just because an OOM notification was received before a crash doesn't necessarily mean this was the cause. To actually allow historical analysis we'll have to do more than just notifications. I'd suggest removing this from the description.

I'd really like to keep this. What if I add the supplementary detail that this doesn't imply historical analysis of time-series resource usage to keep the scope slim? All I'd like to keep in mind is that after a resource alert is triggered, the workspace may fail, and it'd be nice to triage this via notifications, even if that's not strictly a first-class experience.

> Here's my idea for how the monitoring could be implemented on the template-level:

This terraform example helps a ton. Thank you for drafting it. @defelmnq can you please keep me in the loop with the latest changes to this spec, as I am hoping to discuss this with customers ASAP.

**Comment by dannykopping** [comment]
> Both users and administrators want to diagnose whether the agent was killed due to OOM/OOD after crashes occur.

@stirby I understand where you're coming from but this is _related_ but not _required_ to implement OOM/OOD. Knowing whether the agent, specifically, was OOMed is an entirely different piece of work. This will involve OS-specific interventions where we perhaps drop a file somewhere to indicate that the agent was OOM-killed, and then expose that information some way (metric?). I'm not even sure if all platforms offer a way to intercept a process being killed due to excessive memory.

Keep in mind that the OOM/OOD notifications are for the workspace _in general_, not the agent itself, which is why I'm motivating for it not blocking the completion of this task.

I'm in total agreement that we need this, but it's a separate (albeit related) feature - which can be worked on concurrently.

> All I'd like to keep in mind is that after a resource alert is triggered, the workspace may fail, and it'd be nice to triage this via notifications, even if that's not strictly a first-class experience.

This will already be possible since the notification is sent via the control-plane, not the agent, so as long as the agent can send the control-plane the mesurements, an alert can trigger and operators will have something to go on.
